homeassistant/components/rituals_perfume_genie/__init__.py

```diff
@@ -26,12 +29,46 @@
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up Rituals Perfume Genie from a config entry."""
-    session = async_get_clientsession(hass)
-    account = Account(session=session, account_hash=entry.data[ACCOUNT_HASH])
+    _LOGGER.debug("async_setup_entry start (entry_id=%s)", entry.entry_id)
+    session = async_create_clientsession(hass)
+
+    # Require credentials for runtime; if missing, trigger reauth and stop setup
+    if USERNAME not in entry.data or PASSWORD not in entry.data:
+        await _trigger_reauth(hass, entry)
+        return False
+
+    email = entry.data[USERNAME]
+    password = entry.data[PASSWORD]
+
+    # ACCOUNT_HASH is kept only for backwards compatibility
+    account = Account(
+        email=email,
+        password=password,
+        session=session,
+        account_hash=entry.data.get(ACCOUNT_HASH, ""),
+    )
 
     try:
+        # Authenticate first so API token/cookies are available for subsequent calls
+        await account.authenticate()
         account_devices = await account.get_devices()
-    except aiohttp.ClientError as err:
+
+    except AuthenticationException:
+        # Credentials invalid/expired → start reauth and stop setup until user completes it
+        await _trigger_reauth(hass, entry)
+        return False
+
+    except ClientResponseError as err:
+        _LOGGER.warning(
```

Comment:
Do we really need to log this at warning level? There's nothing the user can do, and the set up will be retried automatically. I'd suggest to lower to debug level.